examples/Auction/StateMachine.hs

```diff
@@ -0,0 +1,339 @@
+module Auction.StateMachine (
+  auctionTests,
+  auctionTest,
+) where
+
+import Hedgehog (
+  Callback (Ensure, Require, Update),
+  Command (Command),
+  Concrete (Concrete),
+  FunctorB,
+  Group (Group),
+  MonadGen,
+  Property,
+  PropertyT,
+  Rec (Rec),
+  Symbolic,
+  TraversableB,
+  Var (Var),
+  annotateShow,
+  concrete,
+  executeSequential,
+  failure,
+  forAll,
+  property,
+  (===),
+ )
+import Hedgehog.Internal.Distributive (MonadTransDistributive (distributeT))
+
+import Control.Arrow (second)
+import Control.Monad (guard, liftM2)
+import Control.Monad.State.Strict (StateT, evalStateT, runState, state)
+import Data.Function ((&))
+import Data.Kind (Type)
+import Data.List (sort)
+import Data.Map (Map)
+import Data.Maybe (fromMaybe, isJust, isNothing)
+import GHC.Generics (Generic)
+import Plutus.Model (
+  Mock,
+  Run (Run),
+  adaValue,
+  checkErrors,
+  defaultBabbage,
+  initMock,
+ )
+import PlutusLedgerApi.V1 (PubKeyHash)
+import PlutusLedgerApi.V1.Contexts (TxOutRef)
+
+import Auction.PSM qualified as PSM
+import Data.Map qualified as Map
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Internal.Range qualified as Range
+
+newtype User = User {name :: String}
+  deriving stock (Eq, Ord, Show)
+
+data AuctionState v = AuctionState
+  { currentBid :: Maybe (User, User, Int, Var TxOutRef v)
+  , -- auction owner, current bidder, bid, current utxo
+    winner :: Maybe User
+  , users :: Map User (Int, Var PubKeyHash v)
+  }
+  deriving stock (Eq, Ord, Show)
```

Comment:
This needs to better reflect the state of the contract. It probably need some type of sum type to capture the current state. It also needs to reflect start and end states. Remember, each transaction changes the state.

Reply:
I took out winner which wasn't really doing anything and renamed currentBid to currentAuction. I think Maybe already is the Enum type we want for that because there's no need to distinguish the start and end state and it's fine for a test to have more than one auction take place.